url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/server/SOCChannelList.java

```diff
@@ -0,0 +1,339 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+ *
+ * The author of this program can be reached at thomas@infolab.northwestern.edu
+ **/
+package soc.server;
+
+import soc.disableDebug.D;
+
+import soc.server.genericServer.StringConnection;
+
+import soc.util.MutexFlag;
+
+import java.util.Enumeration;
+import java.util.Hashtable;
+import java.util.Vector;
+
+
+/**
+ * A class for tracking the chat channels.
+ * The list itself, and each channel, has a monitor for synchronization.
+ *
+ * @author Robert S. Thomas
+ */
+public class SOCChannelList
+{
+    /** key = string, value = Vector of MutexFlags */
+    protected Hashtable channelMutexes;
+
+    /** key = string, value = Vector of StringConnections */
+    protected Hashtable channelMembers;
+
+    /** track the monitor for this channel list */
+    protected boolean inUse;
+
+    /**
+     * constructor
+     */
+    public SOCChannelList()
+    {
+        channelMutexes = new Hashtable();
+        channelMembers = new Hashtable();
+        inUse = false;
+    }
+
+    /**
+     * take the monitor for this channel list
+     */
+    public synchronized void takeMonitor()
+    {
+        D.ebugPrintln("SOCChannelList : TAKE MONITOR");
+
+        while (inUse)
+        {
+            try
+            {
+                wait(1000);
+            }
+            catch (InterruptedException e)
+            {
+                System.out.println("EXCEPTION IN takeMonitor() -- " + e);
+            }
+        }
+
+        inUse = true;
+    }
+
+    /**
+     * release the monitor for this channel list
+     */
+    public synchronized void releaseMonitor()
+    {
+        D.ebugPrintln("SOCChannelList : RELEASE MONITOR");
+        inUse = false;
+        this.notify();
+    }
+
+    /**
+     * take the monitor for this channel
+     *
+     * @param channel  the name of the channel
+     * @return false   if the channel has no mutex
+     */
+    public boolean takeMonitorForChannel(String channel)
+    {
+        D.ebugPrintln("SOCChannelList : TAKE MONITOR FOR " + channel);
+
+        MutexFlag mutex = (MutexFlag) channelMutexes.get(channel);
+
+        if (mutex == null)
+        {
+            return false;
+        }
+
+        boolean done = false;
+
+        while (!done)
+        {
+            mutex = (MutexFlag) channelMutexes.get(channel);
+
+            if (mutex == null)
+            {
+                return false;
+            }
+
+            synchronized (mutex)
+            {
+                if (mutex.getState() == true)
+                {
+                    try
+                    {
+                        mutex.wait(1000);
+                    }
+                    catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)